scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala

```diff
@@ -0,0 +1,153 @@
+            val forwardCompatElement = formatForwardCompatibleType(sourceElement, targetElement)
+            // Wrap the solved element with current source structure, and do actual conversion work
+            val forwardCompatRepeated = ParquetListFormatRule.wrapElementAsRepeatedType(
+              sourceRule,
+              sourceRepeatedListType,
+              forwardCompatElement
+            )
+            ParquetListFormatRule.formatForwardCompatibleRepeatedType(
+              forwardCompatRepeated,
+              targetRepeatedListType)
+          }
+          case _ => sourceRepeatedListType // No-op
+        }
+        // Wrapped the formatted repeated type in its original groups,
+        // describing field name and whether it's optional/required
+        sourceWrappers.foldRight(formattedRepeated) {
+          (wrapper, group) => wrapper.withNewFields(group)
+        }
+      case (
+        GroupUnwrapped(sourceWrappers, None, Some(sourceRepeatedMapType)),
+        GroupUnwrapped(_, None, Some(targetRepeatedMapType))
+      ) =>
+        // Format map
+        val forwardCompatRepeated = formatForwardCompatibleType(sourceRepeatedMapType, targetRepeatedMapType)
+        val formattedRepeated = ParquetMapFormatRule.formatForwardCompatibleRepeatedType(
+          forwardCompatRepeated,
+          targetRepeatedMapType
+        )
+        // Wrapped the formatted repeated type in its original groups,
+        // describing field name and whether it's optional/required
+        sourceWrappers.foldRight(formattedRepeated) {
+          (wrapper, group) => wrapper.withNewFields(group)
+        }
+      case _ =>
+        // Field projection
+        val sourceGroup = sourceType.asGroupType
+        val targetGroup = targetType.asGroupType
+
+        import scala.collection.JavaConverters._
+        val resultFields = sourceGroup.getFields.asScala.map { sourceField =>
+          if (!targetGroup.containsField(sourceField.getName)) {
+            if (!sourceField.isRepetition(Repetition.OPTIONAL)) {
+              throw new DecodingSchemaMismatchException(
+                s"Found non-optional source field ${sourceField.getName}:\n$sourceField\n\n" +
+                  s"not present in the given target type:\n${targetGroup}"
+              )
+            }
+            sourceField
+          }
+          else {
+            val fieldIndex = targetGroup.getFieldIndex(sourceField.getName)
+            val targetField = targetGroup.getFields.get(fieldIndex)
+            formatForwardCompatibleType(sourceField, targetField)
+          }
+        }
+        sourceGroup.withNewFields(resultFields.asJava)
+    }
+  }
+
+  private case class GroupUnwrapped(wrappers: Seq[GroupType],
+                                    repeatedListType: Option[Type] = None,
+                                    repeatedMapType: Option[Type] = None)
+
+  private def unwrapGroup(typ: Type, wrappers: Seq[GroupType] = Seq()): GroupUnwrapped = {
```

Comment:
Perhaps a naive question, but what does "wrapped" or "wrapper" refer to in this case? Could you add a comment explaining this? That might make the following code easier to review.

Reply:
I now simplified to just extract for only the repeated type, and let the recursive call in main method `formatForwardCompatibleType` handle multiple layers instead.
This now becomes a sealed trait for the group type itself, and a helper method to get its repeated type
```
private[scrooge] sealed trait CollectionGroup {
  def groupType: GroupType

  def repeatedType: Type
}
```